api/include/opentelemetry/trace/trace_state.h

```diff
@@ -97,37 +107,163 @@ class TraceState
     }
   };
 
-  // An empty TraceState.
-  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
+  /**
+   * Returns a newly created Tracestate parsed from the header provided.
+   * @param header Encoding of the tracestate header defined by
+   * the W3C Trace Context specification https://www.w3.org/TR/trace-context/
+   *  @return Tracestate A new Tracestate instance or DEFAULT
+   */
+  static TraceState FromHeader(nostd::string_view header)
+  {
+    TraceState ts;
 
-  // Returns false if no such key, otherwise returns true and populates the value parameter with the
-  // associated value.
-  bool Get(nostd::string_view key, nostd::string_view &value) const noexcept
+    std::size_t begin{0};
+    std::size_t end{0};
+    bool invalid_header = false;
+    while (begin < header.size() && ts.num_entries_ < kMaxKeyValuePairs)
+    {
+      // find list-member
+      end = header.find(kMembersSeparator, begin);
+      if (end == std::string::npos)
+      {
+        // last list member. `end` points to end of it.
+        end = header.size() - 1;
+      }
+      else
+      {
+        // `end` points to end of current list member
+        end = end - 1;
```

Comment:
Just `--end`? Also could it get `-1` if the first character is `,`?

Reply:
Made it `--end`. If would be fine if first char is `,`, as this will make control to go to below condition:
```
      if (list_member.size() == 0)
      {
        // empty list member, move to next in list
        begin = end + 2;  // begin points to start of next member
        continue;
      }
```
and process next list member. I have added test case to validate this.

Reply:
If the first char is `,`, then `end` will be `-1`, so the call to `TrimString` will trigger access violation when accessing the input string, like `header[static_cast<size_t>(-1)]` which happens before the empty `list_member` check?

Reply:
Yes, you are right. I am fixing and bazel asan fails because of that. I am fixing it now :) Thanks for pointing out the issue.

Reply:
Ok, the easy solution was to add special conditional check for `,` as first char.